modules/rubiconBidAdapter.js

```diff
@@ -949,6 +882,63 @@ function addVideoParameters(data, bidRequest) {
   data.imp[0].video.h = size[1]
 }
 
+function applyFPD(fpd, mediaType, data) {
+  const map = {user: {banner: 'tg_v.', code: 'user'}, context: {banner: 'tg_i.', code: 'site'}, adserver: 'dfp_ad_unit_code'};
+  let obj = {};
+  let keywords = [];
+  const validate = function(e, t) {
+    if (typeof e === 'object' && !Array.isArray(e)) {
+      utils.logWarn('Rubicon: Filtered FPD key: ', t, ': Expected value to be string, integer, or an array of strings/ints');
+    } else if (e) {
+      return (Array.isArray(e)) ? e.filter(value => {
+        if (typeof value !== 'object' && value) return value;
+
+        utils.logWarn('Rubicon: Filtered value: ', value, 'for key', t, ': Expected value to be string, integer, or an array of strings/ints');
+      }).toString() : e.toString();
+    }
+  };
+
+  Object.keys(fpd).filter(value => fpd[value] && map[value] && typeof fpd[value] === 'object').forEach((type) => {
+    obj[map[type].code] = Object.keys(fpd[type]).filter(value => fpd[type][value]).reduce((result, key) => {
+      if (key === 'keywords') {
+        if (!Array.isArray(fpd[type][key]) && mediaType === BANNER) fpd[type][key] = [fpd[type][key]]
+
+        result[key] = fpd[type][key];
+
+        if (mediaType === BANNER) keywords = keywords.concat(fpd[type][key]);
+      } else if (key === 'data') {
+        utils.mergeDeep(result, {ext: {data: fpd[type][key]}});
+      } else if (key === 'adServer' || key === 'pbAdSlot') {
+        (key === 'adServer') ? ['name', 'adSlot'].forEach(name => {
```

Comment:
These go inside of the imp array. Not at top level of the request.

Reply:
Changed to imp[0]